### PR TITLE
Unreal Engine 4 Plugin

### PR DESCRIPTION
**Pull request by YannickLange**
Initial Unreal Engine 4 Plugin for Agones.

The problem with having the plugin within this repository is that people will have to copy the directory in to the Plugins folder of each Unreal Engine project that uses Agones. They would also have to do this when there is a new version of the plugin. In addition to that it might be less clean regarding branching and binaries for different Unreal Engine versions. I made a seperate repository for just the plugin (https://github.com/YannickLange/UEAgones). So people can just clone that. It might be an idea to do "active" development there and then pull the changes to the agones repo? 

Anyway. Try it out, would love to hear feedback!

### CODE REVIEW

**Comment by YannickLange** [comment]
> 🔥 Awwwwweeeeeesome 🔥
> 
> Some thoughts:
> 
> * Should the folder be "Unreal" rather than "UEAgones" ?
> * We'll definitely want some docs for this.
> 
> I also need to find someone who knows Unreal to review 😄

1. I might be better to name the directory Unreal and within that directory Agones, because that directory needs to be copied into the Plugins directory within the Unreal Engine project. Would be weird to have a Plugins > Unreal directory while it is the Agones plugin.
2. I will write those later and give them to you so you can add and test it for the website.

**Comment by think01** [comment]
I'm very interested in trying out this plugin, but I'm not much used to install new plugins into Unreal. Is there some docs I can read/follow to know how to do it right?

**Comment by YannickLange** [comment]
> I'm very interested in trying out this plugin, but I'm not much used to install new plugins into Unreal. Is there some docs I can read/follow to know how to do it right?

I'll write docs later, but it comes down to the general steps. I assume you already have done/been able to do 1, 3, 4 and 6. Only 2 and 4 are related to the plugin. No coding required :smiley:

1. Have project with Unreal Engine from source code 
2. Add this directory to your Plugins directory within your Unreal Engine Project
3. Setup cross compilation so you can compile Unreal Engine + your game (including the plugin) for Linux
4. Compile your game for Linux with Development Server
5. Make sure the plugin is enabled within your game (restart the editor if it wasn't)
6. Package your game

Hope that helps for now!

**Comment by dsazonoff** [comment]
Code looks ok. I don't think that we need to follow google codestyle for that, because UE has its own. Maybe I can build and test a bit later.
A small remark about copyrights in source code: they all should be 2019 - the date when file was added to repo.

**Comment by steven-supersolid** [comment]
I've been wondering how we might approach a Unity SDK so now following this PR with some interest. Has any thought been given to using gRPC instead of using the REST API?
Also, as I don't know Unreal, would it be possible to wrap the current C++ Agones SDK, and would there be value in doing this as that does use generated code?

**Comment by think01** [comment]
> Has any thought been given to using gRPC instead of using the REST API?

I've tried it (on Linux) and others here have done it on Windows. Integration of gRpc in Unreal Engine is quite an hell at the moment, basically for incompatibility between gRpc compiler settings/base libraries and Unreal's.

**Comment by dsazonoff** [comment]
> I've tried it (on Linux) and others here have done it on Windows. Integration of gRpc in Unreal Engine is quite an hell at the moment, basically for incompatibility between gRpc compiler settings/base libraries and Unreal's.
What do you think about using agones-cpp-sdk as a shared library as a base for UE plugin instead of direct implementation of SDK? Shared library will include all necessary gRPC stuff, so there should be no compiler/linker issues.

**Comment by think01** [comment]
> What do you think about using agones-cpp-sdk as a shared library as a base for UE plugin instead of direct implementation of SDK? Shared library will include all necessary gRPC stuff, so there should be no compiler/linker issues.

As I said, I'm far from being an expert, all I know is that I've wasted several weeks trying to build UE4 with gRpc and I ended giving up and resorting to the REST interface.

Having said that, using the shared library *should* solve compiler/linker issues, but my fear is that you could end up finding those same issues between agones-cpp-sdk and Unreal. Unfortunately the Unreal build process is very cumbersome and poor documented: it's very picky on compiler version and flags, I digged through a lot of different problems and errors one after another so I cannot really feel confident on what to say about your question.

I would say Agones is becoming quickly such an important and notable project that someone at Epic could decide to put some effort in helping this integration (at least documenting what's needed).

**Comment by YannickLange** [comment]
> > What do you think about using agones-cpp-sdk as a shared library as a base for UE plugin instead of direct implementation of SDK? Shared library will include all necessary gRPC stuff, so there should be no compiler/linker issues.
> 
> As I said, I'm far from being an expert, all I know is that I've wasted several weeks trying to build UE4 with gRpc and I ended giving up and resorting to the REST interface.
> 
> Having said that, using the shared library _should_ solve compiler/linker issues, but my fear is that you could end up finding those same issues between agones-cpp-sdk and Unreal. Unfortunately the Unreal build process is very cumbersome and poor documented: it's very picky on compiler version and flags, I digged through a lot of different problems and errors one after another so I cannot really feel confident on what to say about your question.
> 
> I would say Agones is becoming quickly such an important and notable project that someone at Epic could decide to put some effort in helping this integration (at least documenting what's needed).
> 
> EDIT: I've found an older thread about this: #138 that ended with this work: https://github.com/iniside/ue-grpc

I worked at Epic and I don't see them integrating Agones at all. They might add a grpc plugin, which would be nice. In addition to that is a plugin way better for the Agones community, so the community has more control over it.

**Comment by YannickLange** [comment]
> Code looks ok. I don't think that we need to follow google codestyle for that, because UE has its own. Maybe I can build and test a bit later.
> A small remark about copyrights in source code: they all should be 2019 - the date when file was added to repo.

1. The codestyle should be Epics/Unreal's C++ code style. That's also what a large percentage of Unreal Engine users are using. 
2. You are right, will modify it

**Comment by Zanderax** [comment]
Awesome work Yannick! Is there a reason that GameServer, WatchGameServer, SetAnnotation and SetLabel aren't implemented?

**Comment by YannickLange** [comment]
> Awesome work Yannick! Is there a reason that GameServer, WatchGameServer, SetAnnotation and SetLabel aren't implemented?

Just haven't had time for that.

**Comment by markmandel** [comment]
Just a gentle bump - our release candidate is hitting tomorrow, at which point we hit feature freeze for a week. Be great to get the outstanding items in this PR done so we can have this in this upcoming release.

**Comment by dsazonoff** [comment]
@YannickLange I suggest to use existing cpp sdk instead of making manual http requests. Now it is possible to build cpp sdk as a dynamic library for all platforms (with CMake and AGONES_BUILD_SHARED=ON), so there will be no direct gRPC dependency (so it is not necessary to build gRPC with UE build system). If you have any questions related to building cpp sdk, feel free to contact me.

File: site/content/en/docs/Guides/Client SDKs/unreal.md [hunk 1]
```diff
@@ -0,0 +1,37 @@
+---
+title: "Unreal Engine Game Server Client Plugin"
+linkTitle: "Unreal Engine"
+date: 2019-01-02T10:17:50Z
```
Comment (markmandel) [inline site/content/en/docs/Guides/Client SDKs/unreal.md]:
```suggestion
date: 2019-06-13T10:17:50Z
publishDate: 2019-06-13
```

Sorry to take so long to drive this though - but this looks great! Can you make this small change so the docs are hidden until the next release, and then I'm happy to approve and push this through!

**Comment by markmandel** [comment]
Wh00t! We have a working build!

So - final, final things.

1. Per [contributor guidelines](https://github.com/GoogleCloudPlatform/agones/blob/master/CONTRIBUTING.md#submitting-code-via-pull-requests) please squash the commits down to a single commit
1. Please sign the CLA. We can't accept the PR until you do so :disappointed: 

Really looking forward to having this in the repository finally though!!!

**Comment by YannickLange** [comment]
> Wh00t! We have a working build!
> 
> So - final, final things.
> 
> 1. Per [contributor guidelines](https://github.com/GoogleCloudPlatform/agones/blob/master/CONTRIBUTING.md#submitting-code-via-pull-requests) please squash the commits down to a single commit
> 2. Please sign the CLA. We can't accept the PR until you do so 😞
> 
> Really looking forward to having this in the repository finally though!!!

Signed it! Let's go !!!!

**Comment by markmandel** [comment]
> 1. Per [contributor guidelines](https://github.com/GoogleCloudPlatform/agones/blob/master/CONTRIBUTING.md#submitting-code-via-pull-requests) please squash the commits down to a single commit

^ sorry, still got one more step. Almost there!

**Comment by markmandel** [comment]
Gentle bump :smile: just need this squashed, and its good to go.

**Pull request merged**